### PR TITLE
[POC] ieee802154_hal: tester PR for #14371

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR is rebased on top of #14371 and add all radios implemented so far for testing the Radio HAL API.
This means:
- A POC implementation for AT86RF2xx (doesn't include subghz stuff) https://github.com/RIOT-OS/RIOT/pull/14655/commits/27ae6c300a541a29a9bc17d843289a76e8953152
- An (almost) ready to merge implementation for nrf52840 https://github.com/RIOT-OS/RIOT/pull/14655/commits/f726f00c600a35a9d8b304b8504e06f54bb0c5ec
- A ready to merge implementation for CC2538 https://github.com/RIOT-OS/RIOT/pull/14655/commits/ebeb69597c3ddadd5c979cf7f852b221e3cbd954
- A test application for testing these 3 radios. https://github.com/RIOT-OS/RIOT/pull/14655/commits/5aebfa93b64ec888503db4fbb7aab84cd00ce309

### Testing procedure
1. Check the implementations of the radio HAL for each radio. Please focus on the `ieee802154_radio_ops_t`implementation
2. Run `tests/ieee802154_hal` test (for any of this radios).

- Test setting channel and TX power with the `config_phy` command:
```
2020-08-12 18:24:49,390 #  config_phy
2020-08-12 18:24:49,393 # Usage: config_phy <channel> <tx_pow>
```

An out of range value for TX power should give an error

- Send data between nodes using the `txtsnd` command.
```
Usage: txtsnd <long_addr> <len>
```
It's possible to get the long address with the `print_addr` command.

- Perform CCa with the `cca` command:
```
2020-08-12 18:28:57,628 #  cca
2020-08-12 18:28:57,629 # CLEAR
> 2020-08-12 18:28:57,827 #  cca
2020-08-12 18:28:57,828 # CLEAR
> 2020-08-12 18:28:58,157 #  cca
2020-08-12 18:28:58,158 # CLEAR
> 2020-08-12 18:28:58,334 #  cca
2020-08-12 18:28:58,335 # BUSY
> 2020-08-12 18:28:58,473 #  cca
2020-08-12 18:28:58,474 # CLEAR
```

- Change the CCA mode and the threshold with the `config_cca` command:
```
config_cca <ed|cs|or|and> [ED threshold]>
```

Note the ED is in dBm. The ED threshold doesn't affect the Carrier Sense mode (CCA mode 2).

- Try different TX modes (CSMA-CA, CCA, direct transmission) using the `tx_mode` command:
```
Usage: tx_mode <csma_ca|cca|direct>
```

- Try different RX states (ACK on, ACK disabled, ACK with Frame Pending or Promiscuous mode) with the `rx_mode` command:
```
Usage: rx_mode <on|off|pend|promisc>
``` 

Use a sniffer to check if the receiver sends ACK. As expected, a radio with Promiscuous Mode will receive all kind of packets with correct FCS.

- Use the `caps` command to get the available caps for a given device:
```
2020-08-13 13:33:37,385 # caps
2020-08-13 13:33:37,388 #                CAPS              
2020-08-13 13:33:37,391 # =================================
2020-08-13 13:33:37,394 # - Frame Retransmissions:        y
2020-08-13 13:33:37,397 #     * Info retries:             y
2020-08-13 13:33:37,400 # - Auto CSMA-CA:                 y
2020-08-13 13:33:37,403 # - 2.4 GHz band:                 y
2020-08-13 13:33:37,406 # - SubGHz band:                  n
2020-08-13 13:33:37,412 # - TX DONE indication:           y
2020-08-13 13:33:37,415 #     * ACK Timeout indication:   n
2020-08-13 13:33:37,418 # - RX_START indication:          n
2020-08-13 13:33:37,420 # - CCA Done indication:          y
```

Example session between a `cc2538dk` and `nrf52840`:

<details><summary>NRF52840:</summary>

```
2020-07-29 18:19:56,350 # Initialization successful - starting the shell now
print_addrli
2020-07-29 18:20:05,050 #  print_addr
2020-07-29 18:20:05,051 # 3a72df8cb6556f1d
> 2020-07-29 18:20:07,537 #  Packet received:
2020-07-29 18:20:07,549 # 61 dc 00 23 00 1d 6f 55 b6 8c df 72 3a 25 2e 42 fe 23 31 68 22 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 2
txtsnd 22683123fe422e25 100
2020-07-29 18:20:10,986 # txtsnd 22683123fe422e25 100
2020-07-29 18:20:10,993 # TX_DONE!
> txtsnd 22683123fe422e2e 100
2020-07-29 18:20:20,533 #  txtsnd 22683123fe422e2e 100
2020-07-29 18:20:20,538 # TX_DONE!
> config_phy 19 14
2020-07-29 18:20:29,788 #  config_phy 19 14
2020-07-29 18:20:29,790 # Success!
> 2020-07-29 18:20:45,803 #  Packet received:
2020-07-29 18:20:45,816 # 61 dc 03 23 00 1d 6f 55 b6 8c df 72 3a 25 2e 42 fe 23 31 68 22 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 2
> 2020-07-29 18:20:53,906 #  Packet received:
2020-07-29 18:20:53,918 # 61 dc 04 23 00 1d 6f 55 b6 8c df 72 3a 25 2e 42 fe 23 31 68 22 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 2
> 2020-07-29 18:21:00,818 #  Packet received:
2020-07-29 18:21:00,831 # 61 dc 05 23 00 1d 6f 55 b6 8c df 72 3a 25 2e 42 fe 23 31 68 22 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 2
> 2020-07-29 18:21:09,375 #  Packet received:
2020-07-29 18:21:09,388 # 61 dc 06 23 00 1d 6f 55 b6 8c df 72 3a 25 2e 42 fe 23 31 68 22 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 2
2020-07-29 18:21:10,038 # Packet received:
2020-07-29 18:21:10,051 # 61 dc 07 23 00 1d 6f 55 b6 8c df 72 3a 25 2e 42 fe 23 31 68 22 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 2
txtsnd 22683123fe422e25 100
2020-07-29 18:21:17,263 # txtsnd 22683123fe422e25 100
2020-07-29 18:21:17,270 # TX_DONE!
> txtsnd 22683123fe422e25 100
2020-07-29 18:21:17,785 #  txtsnd 22683123fe422e25 100
2020-07-29 18:21:17,791 # TX_DONE!
> txtsnd 22683123fe422e25 100
2020-07-29 18:21:18,249 #  txtsnd 22683123fe422e25 100
2020-07-29 18:21:18,255 # TX_DONE!
```
</details>

<details><summary> CC2538DK </summary>

```
2020-07-29 18:17:33,401 # Initialization successful - starting the shell now
print_addr00
2020-07-29 18:17:35,792 #  print_addr
2020-07-29 18:17:35,792 # 22683123fe422e25
txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:19:27,311 #  txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:19:27,316 # TX_DONE!
2020-07-29 18:19:46,682 #  Packet received:
2020-07-29 18:19:46,683 # 61 dc 00 23 00 25 2e 42 fe 23 31 68 22 1d 6f 55 b6 8c df 72 3a LQI: 229, RSSI: 6
2020-07-29 18:19:46,683 # 
2020-07-29 18:19:51,729 # Packet received:
2020-07-29 18:19:51,767 # 61 dc 01 23 00 25 2e 42 fe 23 31 68 22 1d 6f 55 b6 8c df 72 3a 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 20 61 6d 65 74 2c 20 63 6f 6e 73 65 63 74 65 74 75 72 20 61 64 69 70 69 73 63 69 6e 67 20 65 6c 69 74 2e 20 45 74 69 61 6d 20 6f 72 6e 61 72 65 20 6c 61 63 69 6e 69 61 20 6d 69 20 65 6c 65 6d 65 6e 74 75 6d 20 69 6e 74 65 72 64 75 6d 20 LQI: 238, RSSI: 6
2020-07-29 18:19:51,767 # 
2020-07-29 18:19:57,567 # main(): This is RIOT! (Version: 2020.07-devel-1612-gdd739-pr/radio_hal_tester)
2020-07-29 18:19:57,568 # Initialization successful - starting the shell now
print_addr00
2020-07-29 18:20:02,218 #  print_addr
2020-07-29 18:20:02,219 # 22683123fe422e25
txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:20:07,531 #  txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:20:07,536 # TX_DONE!
> 2020-07-29 18:20:10,993 #  Packet received:
2020-07-29 18:20:11,030 # 61 dc 00 23 00 25 2e 42 fe 23 31 68 22 1d 6f 55 b6 8c df 72 3a 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 20 61 6d 65 74 2c 20 63 6f 6e 73 65 63 74 65 74 75 72 20 61 64 69 70 69 73 63 69 6e 67 20 65 6c 69 74 2e 20 45 74 69 61 6d 20 6f 72 6e 61 72 65 20 6c 61 63 69 6e 69 61 20 6d 69 20 65 6c 65 6d 65 6e 74 75 6d 20 69 6e 74 65 72 64 75 6d 20 LQI: 233, RSSI: 7
2020-07-29 18:20:11,030 # 
txtsnd 3a72df8cb6556f1e 100
2020-07-29 18:20:16,934 # txtsnd 3a72df8cb6556f1e 100
2020-07-29 18:20:16,940 # TX_DONE!
txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:20:38,113 #  txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:20:38,118 # TX_DONE!
> config_phy 19 14
2020-07-29 18:20:44,729 #  config_phy 19 14
2020-07-29 18:20:44,730 # Success!
txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:20:45,798 #  txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:20:45,802 # TX_DONE!
> txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:20:53,900 #  txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:20:53,905 # TX_DONE!
> txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:21:00,812 #  txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:21:00,817 # TX_DONE!
> txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:21:09,369 #  txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:21:09,374 # TX_DONE!
> txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:21:10,033 #  txtsnd 3a72df8cb6556f1d 100
2020-07-29 18:21:10,037 # TX_DONE!
> 2020-07-29 18:21:17,270 #  Packet received:
2020-07-29 18:21:17,311 # 61 dc 0b 23 00 25 2e 42 fe 23 31 68 22 1d 6f 55 b6 8c df 72 3a 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 20 61 6d 65 74 2c 20 63 6f 6e 73 65 63 74 65 74 75 72 20 61 64 69 70 69 73 63 69 6e 67 20 65 6c 69 74 2e 20 45 74 69 61 6d 20 6f 72 6e 61 72 65 20 6c 61 63 69 6e 69 61 20 6d 69 20 65 6c 65 6d 65 6e 74 75 6d 20 69 6e 74 65 72 64 75 6d 20 LQI: 238, RSSI: 7
2020-07-29 18:21:17,311 # 
2020-07-29 18:21:17,791 # Packet received:
2020-07-29 18:21:17,828 # 61 dc 0c 23 00 25 2e 42 fe 23 31 68 22 1d 6f 55 b6 8c df 72 3a 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 20 61 6d 65 74 2c 20 63 6f 6e 73 65 63 74 65 74 75 72 20 61 64 69 70 69 73 63 69 6e 67 20 65 6c 69 74 2e 20 45 74 69 61 6d 20 6f 72 6e 61 72 65 20 6c 61 63 69 6e 69 61 20 6d 69 20 65 6c 65 6d 65 6e 74 75 6d 20 69 6e 74 65 72 64 75 6d 20 LQI: 238, RSSI: 8
2020-07-29 18:21:17,829 # 
2020-07-29 18:21:18,256 # Packet received:
2020-07-29 18:21:18,296 # 61 dc 0d 23 00 25 2e 42 fe 23 31 68 22 1d 6f 55 b6 8c df 72 3a 4c 6f 72 65 6d 20 69 70 73 75 6d 20 64 6f 6c 6f 72 20 73 69 74 20 61 6d 65 74 2c 20 63 6f 6e 73 65 63 74 65 74 75 72 20 61 64 69 70 69 73 63 69 6e 67 20 65 6c 69 74 2e 20 45 74 69 61 6d 20 6f 72 6e 61 72 65 20 6c 61 63 69 6e 69 61 20 6d 69 20 65 6c 65 6d 65 6e 74 75 6d 20 69 6e 74 65 72 64 75 6d 20 LQI: 238, RSSI: 7
```
</details>

### Issues/PRs references

#14371
#13943

### CODE REVIEW

**Comment by jia200x** [comment]
@PeterKietzmann @maribu @benpicco if you all agree, I would like to move the CC2538 commits out of #14371 and move them here. That way #14371 only focus on the API. I can keep rebasing this one.

**Comment by jia200x** [comment]
done!

File: tests/ieee802154_hal/Makefile [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ifneq (,$(filter samr21-xpro,$(BOARD)))
+  DRIVER := at86rf233
+endif
+ifneq (,$(filter iotlab-m3 fox,$(BOARD)))
+  DRIVER := at86rf231
+endif
+ifneq (,$(filter nrf52840dk,$(BOARD)))
+  DRIVER := nrf802154
+endif
+ifneq (,$(filter cc2538dk,$(BOARD)))
```
Comment (PeterKietzmann) [inline tests/ieee802154_hal/Makefile:31]: Would be fine to check for the CC2538 MCU here. However, you were talking about the remote-rev% board somewhere, not cc2538dk.

Comment (PeterKietzmann) [inline tests/ieee802154_hal/Makefile:31]: Would be fine to check for the CC2538 MCU here. However, you were talking about the remote-rev% board somewhere, not cc2538dk.

**Comment by PeterKietzmann** [comment]
I think we should specify a test procedure according to which we test all combinations of devices

**Comment by PeterKietzmann** [comment]
The issues reported [here](https://github.com/RIOT-OS/RIOT/pull/14655#pullrequestreview-458258638) are fixed with #14672 and #14673.

File: cpu/nrf52/radio/nrf802154/nrf802154_radio.c [hunk 2]
```diff
@@ -347,6 +349,21 @@ void nrf802154_on(void)
...
 }
 
+static void _timer_cb(void *arg, int chan)
+{
+    (void)arg;
+    (void)chan;
+    ieee802154_dev_t *dev = &nrf802154_hal_dev;
+
+    ack[0] = 5;
+    ack[1] = 0x02;
```
Comment (benpicco) [inline cpu/nrf52/radio/nrf802154/nrf802154_radio.c:359]:
```suggestion
    ack[1] = IEEE802154_FCF_TYPE_ACK;
```

File: cpu/nrf52/radio/nrf802154/nrf802154_radio.c [hunk 3]
```diff
@@ -347,6 +349,21 @@ void nrf802154_on(void)
 
 }
 
+static void _timer_cb(void *arg, int chan)
+{
+    (void)arg;
+    (void)chan;
+    ieee802154_dev_t *dev = &nrf802154_hal_dev;
+
+    ack[0] = 5;
```
Comment (benpicco) [inline cpu/nrf52/radio/nrf802154/nrf802154_radio.c:358]: Do we have defines for that?

File: cpu/nrf52/radio/nrf802154/nrf802154_radio.c [hunk 4]
```diff
@@ -347,6 +349,21 @@ void nrf802154_on(void)
...
+    ack[1] = 0x02;
+    ack[2] = 0;
```
Comment (benpicco) [inline cpu/nrf52/radio/nrf802154/nrf802154_radio.c:360]: We might just initialize that part and never touch it, should always stay constant.

File: cpu/nrf52/radio/nrf802154/nrf802154_radio.c [hunk 5]
```diff
@@ -347,6 +349,21 @@ void nrf802154_on(void)
...
+    ack[3] = rxbuf[3];
```
Comment (benpicco) [inline cpu/nrf52/radio/nrf802154/nrf802154_radio.c:361]: sequence number should be `ack[2]`

**Comment by jia200x** [comment]
updated to latest changes of #14371

File: cpu/nrf52/radio/nrf802154/nrf802154_radio.c (hunk 5 above)
Comment (jia200x) [reply cpu/nrf52/radio/nrf802154/nrf802154_radio.c:361]: yes, this includes frame length. I will add a macro for the magic number

Comment (benpicco) [reply cpu/nrf52/radio/nrf802154/nrf802154_radio.c:361]: `sizeof(ack)` would be enough. There are already maces for the rest.

**Comment by jia200x** [comment]
I had to re-implement the `set_trx_state` function of nrf52840. The radio behaves weird when switching from TX to RX mode without going to DISABLED state (as seen in the [ERRATA, Nº 204: "Switching between TX and RX causes unwanted emissions"](https://infocenter.nordicsemi.com/topic/errata_nRF52840_Rev2/ERR/nRF52840/Rev2/latest/err_840_changelog.html?cp=4_0_1_0_0)).

I also needed to enable Fast RX-TX transition (this means I had to manually calculate IFS times and reject a transmit request if it doesn't meet the IFS criteria)

**Comment by jia200x** [comment]
I implement Auto CSMA in the CC2538 using the strobe processor. This shows how hardware specific features can be used to optimize the HAL implementation.

**Comment by jia200x** [comment]
I did some minor adjustments with the RSSI and TX power function for all radios. Now their RSSI is compatible with the IEEE 802.15.4 definition and all TX power values were revisited to fit the given dBm value.

**Comment by jia200x** [comment]
updated test procedure

**Comment by jia200x** [comment]
Note the new changes to the test app:

- I added a `caps` command that reports all available caps. E.g samr21-xpro:
```
2020-08-13 13:33:37,385 # caps
2020-08-13 13:33:37,388 #                CAPS              
2020-08-13 13:33:37,391 # =================================
2020-08-13 13:33:37,394 # - Frame Retransmissions:        y
2020-08-13 13:33:37,397 #     * Info retries:             y
2020-08-13 13:33:37,400 # - Auto CSMA-CA:                 y
2020-08-13 13:33:37,403 # - 2.4 GHz band:                 y
2020-08-13 13:33:37,406 # - SubGHz band:                  n
2020-08-13 13:33:37,412 # - TX DONE indication:           y
2020-08-13 13:33:37,415 #     * ACK Timeout indication:   n
2020-08-13 13:33:37,418 # - RX_START indication:          n
2020-08-13 13:33:37,420 # - CCA Done indication:          y
> 
```

- The test application is aware of ACKs now, and reports to the user if a valid ACK is received (if device doesn't support frame retransmissions or ACK timeout).
- Added TX information (including retransmission attempts, if available):
```
2020-08-13 13:42:05,957 #  txtsnd 22683123fe422e25 10
2020-08-13 13:42:05,962 # Transmission succeeded
2020-08-13 13:42:05,963 # 
2020-08-13 13:42:05,965 # Retransmission attempts: 0
```

**Comment by benpicco** [comment]
So how do we proceed?
Want to make separate PRs for the individual drivers?

**Comment by jia200x** [comment]
I would say so. The CC2420 is in an ok-ish state (I just need to fix some labels and minor stuff). The nrf52840 is also not so far from being in a mergeable state.

The at86rf2xx might need some love.

**Comment by jia200x** [comment]
Rebased. I will leave this PR open as a reference and open implementations for the radios.

**Comment by jia200x** [comment]
since we already have a couple of radios merged with the new Radio HAL API, I think we can close this one.

**Pull request closed without merging**